example/randomCells2D/src/randomCells2D.cpp

```diff
@@ -0,0 +1,244 @@
+        std::size_t pitchRand) const -> void
+    {
+        auto const idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        auto const linearIdx = alpaka::mapIdx<1u>(idx, extent)[0];
+        auto const memoryLocationIdx = idx[0] * pitchRand + idx[1];
+        TRandEngine engine(42, static_cast<std::uint32_t>(linearIdx));
+        states[memoryLocationIdx] = engine;
+    }
+};
+
+struct RunTimestepKernelSingle
+{
+    template<typename TAcc, typename TExtent>
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const& acc,
+        TExtent const extent,
+        RandomEngineSingle<TAcc>* const states,
+        float* const cells,
+        std::size_t pitchRand,
+        std::size_t pitchOut) const -> void
+    {
+        auto const idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        auto const memoryLocationRandIdx = idx[0] * pitchRand + idx[1];
+        auto const memoryLocationOutIdx = idx[0] * pitchOut + idx[1];
+
+        // Setup generator and distribution.
+        RandomEngineSingle<TAcc> engine(states[memoryLocationRandIdx]);
+        alpaka::rand::UniformReal<float> dist;
+
+        float sum = 0;
+        for(unsigned numCalculations = 0; numCalculations < NUM_CALCULATIONS; ++numCalculations)
+        {
+            sum += dist(engine);
+        }
+        cells[memoryLocationOutIdx] = sum / NUM_CALCULATIONS;
+        states[memoryLocationRandIdx] = engine;
+    }
+};
+
+struct RunTimestepKernelVector
+{
+    template<typename TAcc, typename TExtent>
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const& acc,
+        TExtent const extent,
+        RandomEngineVector<TAcc>* const states,
+        float* const cells,
+        std::size_t pitchRand,
+        std::size_t pitchOut) const -> void
+    {
+        auto const idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        auto const memoryLocationRandIdx = idx[0] * pitchRand + idx[1];
+        auto const memoryLocationOutIdx = idx[0] * pitchOut + idx[1];
+
+        // Setup generator and distribution.
+        RandomEngineVector<TAcc> engine(states[memoryLocationRandIdx]); // Load the state of the random engine
+        using DistributionResult =
+            typename RandomEngineVector<TAcc>::ResultContainer<float>; // Container type which will store the
+                                                                       // distribution results
+        unsigned constexpr resultVectorSize = std::tuple_size<DistributionResult>::value; // Size of the result vector
+        alpaka::rand::UniformReal<DistributionResult> dist; // Vector-aware distribution function
+
+
+        float sum = 0;
+        static_assert(NUM_CALCULATIONS % resultVectorSize == 0);
+        for(unsigned numCalculations = 0; numCalculations < NUM_CALCULATIONS / resultVectorSize; ++numCalculations)
+        {
+            auto result = dist(engine);
+            for(unsigned i = 0; i < resultVectorSize; ++i)
+            {
+                sum += result[i];
+            }
```

Comment:
My dream interface:

```suggestion
            for(auto result : dist(engine))
            {
                sum += result;
            }
```

Would need `begin`/`end` for your array types.

Reply:
Agreed, but as you said, the PR is already quite long as it is. Introducing the uint4 et. al. array wrappers without the generator facility makes little sense as they are not currently used anywhere. On the other hand, this PR needs the wrappers in order to work on CUDA. I would opt for merging this PR first, and improving the wrappers later.